packages/core/strapi/lib/types/core/attributes/utils.d.ts

```diff
@@ -0,0 +1,74 @@
+import { Get, KeysBy, NeverGuard, PickBy, SchemaUID } from '../../utils';
+import {
+  Attribute,
+  Type,
+  GetBigIntegerAttributeValue,
+  GetComponentAttributeValue,
+  GetBooleanAttributeValue,
+  GetDecimalAttributeValue,
+  GetDynamicZoneAttributeValue,
+  GetEnumerationAttributeValue,
+  GetFloatAttributeValue,
+  GetIntegerAttributeValue,
+  GetJsonAttributeValue,
+  GetMediaAttributeValue,
+  GetPasswordAttributeValue,
+  GetRelationAttributeValue,
+  GetRichTextAttributeValue,
+  GetStringAttributeValue,
+  GetTextAttributeValue,
+  GetUIDAttributeValue,
+} from '.';
+
+export type PickTypes<T extends Type> = T;
+
+export type GetAttributesKeysByType<TUID extends SchemaUID, TType extends Type, TRule = never> = KeysBy<
+  GetAttributes<TUID>,
+  Attribute<TType> & NeverGuard<TRule, unknown>
+>;
+
+export type GetAttributesByType<TUID extends SchemaUID, TType extends Type, TRule = never> = PickBy<
+  GetAttributes<TUID>,
+  Attribute<TType> & NeverGuard<TRule, unknown>
+>;
+
+export type GetAttribute<T extends SchemaUID, U extends GetAttributesKey<T>> = Get<GetAttributes<T>, U>;
+
+export type GetAttributes<T extends SchemaUID> = Get<Strapi.Schemas[T], 'attributes'>;
+
+export type GetAttributesKey<T extends SchemaUID> = keyof GetAttributes<T>;
+
+export type GetAttributeValue<T extends Attribute> =
+  | GetBigIntegerAttributeValue<T>
+  | GetBooleanAttributeValue<T>
+  | GetComponentAttributeValue<T>
+  | GetDecimalAttributeValue<T>
+  | GetDynamicZoneAttributeValue<T>
+  | GetEnumerationAttributeValue<T>
+  | GetFloatAttributeValue<T>
+  | GetIntegerAttributeValue<T>
+  | GetJsonAttributeValue<T>
+  | GetMediaAttributeValue<T>
+  | GetPasswordAttributeValue<T>
+  | GetRelationAttributeValue<T>
+  | GetRichTextAttributeValue<T>
+  | GetStringAttributeValue<T>
+  | GetTextAttributeValue<T>
+  | GetUIDAttributeValue<T>;
+
+export type GetAttributeValueByKey<T extends SchemaUID, U extends GetAttributesKey<T>> = GetAttribute<T,  U
+> extends infer P extends Attribute ? GetAttributeValue<P>:never;
```

Comment:
this line doesn't seem to be an understandable syntax for Typescript it's also causing an error during compilation.
it would work as such `extends infer P  ? GetAttributeValue<P>:never;` but I am not sure what we are trying to achieve here though

Reply:
I think you've tested this branch in your local VSCode instance, right? If this is the case, then this is why it's not recognized as valid Typescript code.

The `extends` constraints on `infer` type variables is a feature introduced by Typescript 4.7 (https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/) which has been released last week. VSCode didn't release a compatible version yet but it should be out in early June (https://github.com/microsoft/vscode/pull/150313). In the meantime I used the VSCode Insider build to develop the feature.

You can see in the following screenshot that it's actually a valid syntax:
![image](https://user-images.githubusercontent.com/25851739/171019228-8dd84dc1-251c-4354-8c79-9417699bab49.png)

I'll add a warning about that to the PR description though. Thanks for pointing that out. 

As for what it achieves, well it actually prevents using a double-ternary expression to both infer a type from an expression _and_ check if the type extends another one (so that the compiler can narrow it down).
Without it I would've needed to do:

```ts
export type GetAttributeValueByKey<T extends SchemaUID, U extends GetAttributesKey<T>> = GetAttribute<T, U> extends infer P ? P extends Attribute ? GetAttributeValue<P> : never : never;
```

Reply:
yep if that's the case then great, but we'll need to update our typescript package it's currently `4.6.2`, basically our compiler can't compile it either I suppose

Reply:
I did it on my local env but didn't push it :clown_face:
Anyway, I'll revert the 4.7 changes here and put them in another PR that we'll merge later when the ecosystem will be ready.